packages/puppeteer-core/src/api/Page.ts

```diff
@@ -1699,34 +1710,44 @@ export abstract class Page extends EventEmitter<PageEvents> {
   }
 
   /**
-   * @param options - Optional waiting parameters
-   * @returns Promise which resolves when network is idle
+   * Waits for the network to be idle.
+   *
+   * @param options - Options to configure waiting behavior.
+   * @returns A promise which resolves once the network is idle.
    */
-  abstract waitForNetworkIdle(options?: {
-    idleTime?: number;
-    timeout?: number;
-  }): Promise<void>;
+  waitForNetworkIdle(options: WaitForNetworkIdleOptions = {}): Promise<void> {
+    return firstValueFrom(this.waitForNetworkIdle$(options));
+  }
 
   /**
    * @internal
    */
-  _waitForNetworkIdle(
-    networkManager: BidiNetworkManager | CdpNetworkManager,
-    idleTime: number,
-    requestsInFlight = 0
+  waitForNetworkIdle$(
+    options: WaitForNetworkIdleOptions = {}
   ): Observable<void> {
+    const {
+      timeout: ms = this._timeoutSettings.timeout(),
+      idleTime = NETWORK_IDLE_TIME,
+      concurrency = 0,
+    } = options;
+
     return merge(
-      fromEmitterEvent(networkManager, NetworkManagerEvent.Request),
-      fromEmitterEvent(networkManager, NetworkManagerEvent.Response),
-      fromEmitterEvent(networkManager, NetworkManagerEvent.RequestFailed)
+      fromEmitterEvent(this, PageEvent.Request),
+      fromEmitterEvent(this, PageEvent.Response),
+      fromEmitterEvent(this, PageEvent.RequestFailed)
     ).pipe(
       startWith(undefined),
-      filter(() => {
-        return networkManager.inFlightRequestsCount() <= requestsInFlight;
-      }),
-      switchMap(() => {
-        return of(undefined).pipe(delay(idleTime));
-      })
+      bufferCount(concurrency + 1, 1),
+      debounceTime(idleTime),
```

Comment:
I think we don't have a proper test, but we never check if the queue is empty or full. 
This will mean that we can get a single Request event that never resolves and this will still emit.

Reply:
Discussed offline.